packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java

```diff
@@ -0,0 +1,248 @@
+package io.flutter.plugins.imagepicker;
+
+import android.Manifest;
+import android.app.Activity;
+import android.content.Intent;
+import android.content.pm.PackageManager;
+import android.content.pm.ResolveInfo;
+import android.media.MediaScannerConnection;
+import android.net.Uri;
+import android.os.Environment;
+import android.provider.MediaStore;
+import android.support.v4.app.ActivityCompat;
+import io.flutter.plugin.common.MethodCall;
+import io.flutter.plugin.common.MethodChannel;
+import io.flutter.plugin.common.PluginRegistry;
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.UUID;
+
+public class ImagePickerDelegate
+    implements PluginRegistry.ActivityResultListener,
+        PluginRegistry.RequestPermissionsResultListener {
+  private static final int REQUEST_CODE_CHOOSE_FROM_GALLERY = 2342;
+  private static final int REQUEST_CODE_TAKE_WITH_CAMERA = 2343;
+  private static final int REQUEST_IMAGE_PICKER_PERMISSIONS = 2344;
+
+  private final Activity registrar;
+  private final ImageResizer imageResizer;
+  private final String providerName;
+
+  private Uri pendingCameraImageUri;
+  private MethodChannel.Result pendingResult;
+  private MethodCall methodCall;
+  private ImagePickingOperationType pendingPickImageOperation;
+
+  public static class FileProvider extends android.support.v4.content.FileProvider {}
+
+  private enum ImagePickingOperationType {
+    CHOOSE_FROM_GALLERY,
+    TAKE_WITH_CAMERA
+  }
+
+  public ImagePickerDelegate(Activity registrar, ImageResizer imageResizer) {
+    this.registrar = registrar;
+    this.imageResizer = imageResizer;
+    this.providerName = registrar.getPackageName() + ".flutter.image_provider";
+  }
+
+  public void chooseImageFromGallery(MethodCall methodCall, MethodChannel.Result result) {
+    setPendingMethodCallAndResult(methodCall, result);
+
+    if (!hasRequiredPermissions()) {
```

Comment:
Do you also need the Camera permission to pick an image from gallery? If not, can we only request the permissions that you actually need to perform the requested operation?